### PR TITLE
Feat/article flatlist

### PR DESCRIPTION
**Pull request by sericaia**
This is work in progress.

FlatList is being used for ios and android and ListView for web.

image

### CODE REVIEW

File: packages/article/article-content.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  <FlatList
+    testID="scroll-view-article"
+    keyExtractor={item => item.type + item.index || item.type}
+    data={data}
+    renderItem={({ item }) => renderRow(item)}
+  />
+);
+
+ArticleContent.propTypes = {
+  data: PropTypes.arrayOf(PropTypes.object),
```
Comment (sericaia) [inline packages/article/article-content.js]: .isRequired?

File: packages/article/article-content.web.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    renderRow={renderRow}
+    initialListSize={initialListSize}
+    scrollRenderAheadDistance={scrollRenderAheadDistance}
+    pageSize={pageSize}
+    enableEmptySections
+  />
+);
+
+ArticleContent.propTypes = {
+  data: PropTypes.arrayOf(PropTypes.object),
```
Comment (sericaia) [inline packages/article/article-content.web.js]: .isRequired?

File: packages/article/article-content.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  <FlatList
+    testID="flat-list-article"
+    keyExtractor={item => item.type + item.index || item.type}
+    data={data}
+    renderItem={({ item }) => renderRow(item)}
+  />
+);
+
+ArticleContent.propTypes = {
+  data: PropTypes.arrayOf(PropTypes.object).isRequired,
```
Comment (jimhunty) [inline packages/article/article-content.js]: Can we add a shape of the object here?

File: packages/article/article.fructose.ios.js [hunk 4]
```diff
@@ -21,7 +21,9 @@ withComponent(
     });
 
     test("default Article should render correctly", async () => {
-      await expect(element(by.id("listView"))).toBeVisible();
+      await expect(
```
Comment (jimhunty) [inline packages/article/article.fructose.ios.js]: Async await seem redundant in this test.

File: packages/article/article-content.web.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { ScrollView } from "react-native";
+
+const ArticleContent = ({ data, renderRow }) => (
+  <ScrollView testID="scroll-view-article">
+    {data.map(item => renderRow(item))}
+  </ScrollView>
+);
+
+ArticleContent.propTypes = {
+  data: PropTypes.arrayOf(PropTypes.object).isRequired,
```
Comment (jimhunty) [inline packages/article/article-content.web.js]: Again it would be useful to have a shape here

File: packages/article/article.js [hunk 6]
```diff
@@ -29,14 +29,15 @@ class ArticlePage extends React.Component {
   static renderRow(rowData) {
     if (rowData.type === "leadAsset") {
       return (
-        <View testID="leadAsset" style={styles.leadAsset}>
+        <View testID="leadAsset" key={rowData.type} style={styles.leadAsset}>
```
Comment (jimhunty) [inline packages/article/article.js]: Is the type guaranteed to be unique?

Comment (sericaia) [reply packages/article/article.js]: yes, there's always only one leadAsset

File: packages/article/article.js [hunk 7]
```diff
@@ -29,14 +29,15 @@ class ArticlePage extends React.Component {
...
           <Image uri={rowData.data.crop.url} aspectRatio={16 / 9} />
```
[the remark this answers is not in the document]
Comment (sericaia) [reply packages/article/article.js]: @nehasri89

File: packages/article/article.js [hunk 8]
```diff
@@ -28,9 +28,11 @@ const withAdComposer = (children, section = "article") => (
 class ArticlePage extends React.Component {
   static renderRow(rowData) {
     if (rowData.type === "leadAsset") {
+      const [ratioWidth, ratioHeight] = rowData.data.crop.ratio.split(":");
```
Comment (craigbilner) [inline packages/article/article.js]: does this need to happen every row?

Comment (killanaca) [reply packages/article/article.js]: it's not every row, it's just for the row of type `leadAsset` (one per article). Maybe at some point we can have some utils package (like the `jest-configurator`)  where we can share this kind of code.

**Pull request merged**